src/MPoly.jl

```diff
@@ -864,21 +864,18 @@ defined between elements of the coefficient ring of $a$ and elements of the
 supplied vector.
 """
 function evaluate(a::MPolyRingElem{T}, vals::Vector{U}) where {T <: RingElement, U <: RingElement}
-   length(vals) != nvars(parent(a)) && error("Incorrect number of values in evaluation")
+   @req length(vals) == nvars(parent(a)) "Incorrect number of values in evaluation"
    R = base_ring(a)
-   if (U <: Integer && U != BigInt) ||
-      (U <: Rational && U != Rational{BigInt})
+   if (U <: Integer && U !== BigInt) ||
+      (U <: Rational && U !== Rational{BigInt})
       c = zero(R)*zero(U)
       V = typeof(c)
-      if U != V
-         vals = [parent(c)(v) for v in vals]
-         powers = [Dict{Int, V}() for i in 1:length(vals)]
-      else
-         powers = [Dict{Int, U}() for i in 1:length(vals)]
+      @assert V == elem_type(R)
```

Comment:
In general, this is wrong (as the failing test shows). With Base types, there might be promotion to a strictly larger type:
```
julia> big"2" * 1//2 |> typeof
Rational{BigInt}
```
So evaluating an `MPoly{BigInt}` at `Rational{Int}` will hit this assertion.

(I remember the pain to try to accommodate this promotion madness. At some point this needs to be overhauled, dealing with the Base types better.)

Reply:
Thanks for the hint much appreciated! I didn't understand the code and therefore decided to put in the `@assert`, thinking it will either work, or else reveal a problematic case (in hindsight I should then also have marked the PR as "draft" but luckily the tests *do* catch it).

I plan to adjust this and then add in a comment explaining this, with the hope that it will help a little bit towards a future overhaul.

But I first need to work on a few other things in the next couple days, I am afraid.

Reply:
Yes, looks good, thanks